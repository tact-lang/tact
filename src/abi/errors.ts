// Errors 0-127 are reserved for VM errors
// Errors 128-255 are reserved for contract errors
export const contractErrors = {
    null: { id: 128, message: "Null reference exception" },
    invalidPrefix: { id: 129, message: "Invalid serialization prefix" },
    invalidMessage: { id: 130, message: "Invalid incoming message" },
    constraintsError: { id: 131, message: "Constraints error" },
    accessDenied: { id: 132, message: "Access denied" },
    contractStopped: { id: 133, message: "Contract stopped" },
    invalidArgument: { id: 134, message: "Invalid argument" },
    codeNotFound: { id: 135, message: "Code of a contract was not found" },
    invalidAddress: { id: 136, message: "Invalid address" },
    masterchainNotEnabled: {
        id: 137,
        message: "Masterchain support is not enabled for this contract",
    },
};
