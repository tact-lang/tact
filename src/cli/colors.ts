import { release } from "os";

export const isColorSupported = () => {
    if (process.env.CI) {
        return true;
    }
    if (process.platform === "win32") {
        const [major, _, build] = release().split(".").map(Number);
        // Windows 10, Build 10586+
        return (
            (major && major > 10) || (major === 10 && build && build >= 10586)
        );
    }
    if (process.stdout.isTTY && process.env.TERM !== "dumb") {
        return true;
    }
    return false;
};

const format = (code: number) => (s: string) => `\x1b[${code}m${s}\x1b[39m`;

const cond =
    (isEnabled: boolean) => (f: (x: string) => string) => (x: string) =>
        isEnabled ? f(x) : x;

export const getAnsiMarkup = (isEnabled: boolean, f = cond(isEnabled)) => ({
    reset: f((s) => `\x1b[0m${s}`),
    bold: f((s) => `\x1b[1m${s}\x1b[22m`),
    red: f(format(31)),
    green: f(format(32)),
    yellow: f(format(33)),
    blue: f(format(34)),
    magenta: f(format(35)),
    cyan: f(format(36)),
    white: f(format(37)),
    gray: f(format(90)),
});

export type AnsiMarkup = ReturnType<typeof getAnsiMarkup>;
