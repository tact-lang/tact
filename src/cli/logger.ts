export const CliLogger = () => {
    let hadErrors = false;

    const log = (message: string) => {
        hadErrors = true;
        console.log(message);
    };

    return {
        log,
        hadErrors: () => hadErrors,
    };
};
