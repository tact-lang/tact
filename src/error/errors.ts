import path from "path";
import { cwd } from "process";
import { AstFuncId, AstId, AstTypeId } from "../ast/ast";
import { SrcInfo } from "../grammar";

export class TactError extends Error {
    readonly loc?: SrcInfo;
    constructor(message: string, loc?: SrcInfo) {
        super(message);
        this.loc = loc;
    }
}

// Any regular compilation error shown to user:
// parsing, typechecking, code generation
export class TactCompilationError extends TactError {
    constructor(message: string, loc?: SrcInfo) {
        super(message, loc);
    }
}

export class TactInternalCompilerError extends TactError {
    constructor(message: string, loc?: SrcInfo) {
        super(message, loc);
    }
}

export class TactConstEvalError extends TactCompilationError {
    fatal: boolean = false;
    constructor(message: string, fatal: boolean, loc: SrcInfo) {
        super(message, loc);
        this.fatal = fatal;
    }
}

export function locationStr(sourceInfo: SrcInfo): string {
    if (sourceInfo.file) {
        const loc = sourceInfo.interval.getLineAndColumn() as {
            lineNum: number;
            colNum: number;
        };
        const file = path.relative(cwd(), sourceInfo.file);
        return `${file}:${loc.lineNum}:${loc.colNum}: `;
    } else {
        return "";
    }
}

export function throwCompilationError(
    message: string,
    source?: SrcInfo,
): never {
    const msg =
        source === undefined
            ? message
            : `${locationStr(source)}${message}\n${source.interval.getLineAndColumnMessage()}`;
    throw new TactCompilationError(msg, source);
}

export function throwInternalCompilerError(
    message: string,
    source?: SrcInfo,
): never {
    const msg = `[INTERNAL COMPILER ERROR]: ${message}\nPlease report at https://github.com/tact-lang/tact/issues`;
    throw source === undefined
        ? new TactInternalCompilerError(msg)
        : new TactInternalCompilerError(
              `${locationStr(source)}\n${msg}\n${source.interval.getLineAndColumnMessage()}`,
              source,
          );
}

export function throwConstEvalError(
    message: string,
    fatal: boolean,
    source: SrcInfo,
): never {
    throw new TactConstEvalError(
        `${locationStr(source)}${message}\n${source.interval.getLineAndColumnMessage()}`,
        fatal,
        source,
    );
}

export function idTextErr(
    ident: AstId | AstFuncId | AstTypeId | string,
): string {
    if (typeof ident === "string") {
        return `"${ident}"`;
    }
    return `"${ident.text}"`;
}

export type TactErrorCollection =
    | Error
    | TactCompilationError
    | TactInternalCompilerError
    | TactConstEvalError;
