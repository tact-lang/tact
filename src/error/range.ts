/**
 * Range of source code positions
 * 0-based
 * `end` points to position just after the end of range
 */
export type Range = {
    start: number;
    end: number;
};

/**
 * Intersect ranges
 */
export const intersect = (a: Range, b: Range): Range => {
    return {
        start: Math.max(a.start, b.start),
        end: Math.min(a.end, b.end),
    };
};

/**
 * Shift both sides of range by scalar
 */
export const shift = (a: Range, b: number) => {
    return {
        start: a.start + b,
        end: a.end + b,
    };
};
