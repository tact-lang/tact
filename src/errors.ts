import { MatchResult } from "ohm-js";
import path from "path";
import { cwd } from "process";
import { ASTRef } from "./grammar/ast";

export class TactError extends Error {
    readonly ref: ASTRef;
    constructor(message: string, ref: ASTRef) {
        super(message);
        this.ref = ref;
    }
}

export class TactParseError extends TactError {
    constructor(message: string, ref: ASTRef) {
        super(message, ref);
    }
}

/// This will be split at least into two categories: typechecking and codegen errors
export class TactCompilationError extends TactError {
    constructor(message: string, ref: ASTRef) {
        super(message, ref);
    }
}

export class TactConstEvalError extends TactCompilationError {
    fatal: boolean = false;
    constructor(message: string, fatal: boolean, ref: ASTRef) {
        super(message, ref);
        this.fatal = fatal;
    }
}

function locationStr(sourceInfo: ASTRef): string {
    if (sourceInfo.file) {
        const loc = sourceInfo.interval.getLineAndColumn() as {
            lineNum: number;
            colNum: number;
        };
        const file = path.relative(cwd(), sourceInfo.file);
        return `${file}:${loc.lineNum}:${loc.colNum}: `;
    } else {
        return "";
    }
}

export function throwParseError(matchResult: MatchResult, path: string): never {
    const interval = matchResult.getInterval();
    const source = new ASTRef(interval, path);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const message = `Parse error: expected ${(matchResult as any).getExpectedText()}\n`;
    throw new TactParseError(
        `${locationStr(source)}${message}\n${interval.getLineAndColumnMessage()}`,
        source,
    );
}

export function throwCompilationError(message: string, source: ASTRef): never {
    throw new TactCompilationError(
        `${locationStr(source)}${message}\n${source.interval.getLineAndColumnMessage()}`,
        source,
    );
}

export function throwConstEvalError(
    message: string,
    fatal: boolean,
    source: ASTRef,
): never {
    throw new TactConstEvalError(
        `${locationStr(source)}${message}\n${source.interval.getLineAndColumnMessage()}`,
        fatal,
        source,
    );
}
