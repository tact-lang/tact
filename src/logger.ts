export type TactLogger = {
    log: (message: string) => void;
    error: (message: string) => void;
};

export const consoleLogger: TactLogger = {
    log: console.log,
    error: console.error,
};
