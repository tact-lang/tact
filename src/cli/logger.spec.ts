/* eslint-disable @typescript-eslint/no-confusing-void-expression */
import { throwInternal } from "../error/errors";
import { Logger, SourceLogger } from "../error/logger-util";
import { Range } from "../error/range";
import { getAnsiMarkup } from "./colors";
import { TerminalLogger } from "./logger";
import pathWindows from "path/win32";
import pathPosix from "path/posix";

const catchProcessExit = <T>(fn: () => T): T | string => {
    const exitSpy = jest.spyOn(process, "exit").mockImplementation((code) => {
        throw new Error(`process.exit called with code ${code}`);
    });

    let result: T | null = null;
    let caughtError: unknown = null;

    try {
        result = fn();
    } catch (error) {
        caughtError = error;
    }

    exitSpy.mockRestore();

    if (caughtError) {
        return (caughtError as Error).message.includes("process.exit called")
            ? (caughtError as Error).message
            : "Unknown error";
    }

    return result as T;
};

const os = [
    ["Windows", pathWindows],
    ["POSIX", pathPosix],
] as const;

describe.each(os)("TerminalLogger %s", (_, pathApi) => {
    const ansi = getAnsiMarkup(false);

    test("only first error without error recovery", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.error(log.text`Error 1`);
                log.error(log.text`Error 2`);
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        logSpy.mockRestore();
    });

    test("all info logs are logged", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.info(log.text`Info 1`);
                log.info(log.text`Info 2`);
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        logSpy.mockRestore();
    });

    test("warn verbosity does not show info", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "warn", ansi, (log) => {
                log.warn(log.text`Warn`);
                log.info(log.text`Info`);
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        logSpy.mockRestore();
    });

    test("path is resolved relative to cwd", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const cwdSpy = jest.spyOn(process, "cwd").mockReturnValue("/foo");
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.error(log.text`See ${log.path("/foo/bar")}.`);
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        cwdSpy.mockRestore();
        logSpy.mockRestore();
    });

    test("raw internal error", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (_log) => {
                throwInternal(`OMG`);
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        logSpy.mockRestore();
    });

    test("logger internal error", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.internal(log.text`OMG`);
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        logSpy.mockRestore();
    });

    test("raw internal error in source", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const cwdSpy = jest.spyOn(process, "cwd").mockReturnValue("/foo");
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.source("/foo/bar", "Hello, world", () => {
                    throwInternal(`OMG`);
                });
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        cwdSpy.mockRestore();
        logSpy.mockRestore();
    });

    test("logger internal error in source", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const cwdSpy = jest.spyOn(process, "cwd").mockReturnValue("/foo");
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.source("/foo/bar", "Hello, world", (log) => {
                    log.internal(log.text`OMG`);
                    log.info(log.text`Impossible`);
                });
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        cwdSpy.mockRestore();
        logSpy.mockRestore();
    });

    test("internal error in source at range", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const cwdSpy = jest.spyOn(process, "cwd").mockReturnValue("/foo");
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.source("/foo/bar", "Hello, world", (log) => {
                    log.at({ start: 3, end: 5 }).internal(log.text`OMG`);
                    log.info(log.text`Impossible`);
                });
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        cwdSpy.mockRestore();
        logSpy.mockRestore();
    });

    test("uncaught error", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, () => {
                throw new Error("Uncaught!");
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        logSpy.mockRestore();
    });

    test("uncaught error in source", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const cwdSpy = jest.spyOn(process, "cwd").mockReturnValue("/foo");
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                return log.source("/foo/bar", "Hello, world", () => {
                    throw new Error("hehe");
                });
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        cwdSpy.mockRestore();
        logSpy.mockRestore();
    });

    test("multiple errors", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.recover((log) => {
                    log.error(log.text`foo`);
                    log.error(log.text`bar`);
                });
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        logSpy.mockRestore();
    });

    test("exit on error", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.recover((log) => {
                    log.error(log.text`foo`);
                    log.error(log.text`bar`);
                    log.exitIfErrored();
                });
                log.error(log.text`impossible`);
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        logSpy.mockRestore();
    });

    test("multiple errors inside source", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const cwdSpy = jest.spyOn(process, "cwd").mockReturnValue("/foo");
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.source("/foo/bar", "Hello, world", (log) => {
                    log.recover((log) => {
                        log.at({ start: 3, end: 5 }).error(log.text`foo`);
                        log.error(log.text`bar`);
                        log.exitIfErrored();
                    });
                    log.error(log.text`impossible`);
                });
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        cwdSpy.mockRestore();
        logSpy.mockRestore();
    });

    test("source inside multiple errors", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const cwdSpy = jest.spyOn(process, "cwd").mockReturnValue("/foo");
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.recover((log) => {
                    log.source("/foo/bar", "Hello, world", (log) => {
                        log.at({ start: 3, end: 5 }).error(log.text`foo`);
                        log.error(log.text`bar`);
                        log.exitIfErrored();
                    });
                    log.error(log.text`impossible`);
                });
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        cwdSpy.mockRestore();
        logSpy.mockRestore();
    });

    test("typed errors", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const cwdSpy = jest.spyOn(process, "cwd").mockReturnValue("/foo");
        const fooBarSchema = <M, R>(l: Logger<M, R>) => ({
            fooError: () => l.error(l.text`Foo!`),
            barError: () => l.error(l.text`Bar!`),
        });
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.recover((log) => {
                    const l = fooBarSchema(log);
                    l.fooError();
                    l.barError();
                });
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        cwdSpy.mockRestore();
        logSpy.mockRestore();
    });

    test("typed errors for source", () => {
        const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});
        const cwdSpy = jest.spyOn(process, "cwd").mockReturnValue("/foo");
        const fooBarSchemaSrc = <M, R>(l: SourceLogger<M, R>) => ({
            fooError: (at: Range) => l.at(at).error(l.text`Foo!`),
            barError: () => l.error(l.text`Bar!`),
        });
        const result = catchProcessExit(() => {
            return TerminalLogger(pathApi, "info", ansi, (log) => {
                log.source("/foo/bar", "Hello, world", (log) => {
                    log.recover((log) => {
                        const l = fooBarSchemaSrc(log);
                        l.fooError({ start: 3, end: 5 });
                        l.barError();
                    });
                });
            });
        });
        expect(result).toMatchSnapshot();
        expect(logSpy.mock.calls).toMatchSnapshot();
        cwdSpy.mockRestore();
        logSpy.mockRestore();
    });
});
