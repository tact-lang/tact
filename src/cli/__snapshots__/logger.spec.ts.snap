// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TerminalLogger POSIX all info logs are logged 1`] = `undefined`;

exports[`TerminalLogger POSIX all info logs are logged 2`] = `
[
  [
    "Info 1",
  ],
  [
    "Info 2",
  ],
]
`;

exports[`TerminalLogger POSIX exit on error 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX exit on error 2`] = `
[
  [
    "Error: foo",
  ],
  [
    "Error: bar",
  ],
]
`;

exports[`TerminalLogger POSIX internal error in source at range 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX internal error in source at range 2`] = `
[
  [
    "Internal compiler error: ./bar:1:4 - OMG
> 1 | Hello, world
         ^~
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger POSIX logger internal error 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX logger internal error 2`] = `
[
  [
    "Internal compiler error: OMG
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger POSIX logger internal error in source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX logger internal error in source 2`] = `
[
  [
    "Internal compiler error: ./bar: OMG
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger POSIX multiple errors 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX multiple errors 2`] = `
[
  [
    "Error: foo",
  ],
  [
    "Error: bar",
  ],
]
`;

exports[`TerminalLogger POSIX multiple errors inside source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX multiple errors inside source 2`] = `
[
  [
    "Error: ./bar:1:4 - foo
> 1 | Hello, world
         ^~",
  ],
  [
    "Error: ./bar: bar",
  ],
]
`;

exports[`TerminalLogger POSIX only first error without error recovery 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX only first error without error recovery 2`] = `
[
  [
    "Error: Error 1",
  ],
]
`;

exports[`TerminalLogger POSIX path is resolved relative to cwd 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX path is resolved relative to cwd 2`] = `
[
  [
    "Error: See ./bar.",
  ],
]
`;

exports[`TerminalLogger POSIX raw internal error 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX raw internal error 2`] = `
[
  [
    "Internal compiler error: OMG
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger POSIX raw internal error in source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX raw internal error in source 2`] = `
[
  [
    "Internal compiler error: OMG
while compiling /foo/bar
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger POSIX source inside multiple errors 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX source inside multiple errors 2`] = `
[
  [
    "Error: ./bar:1:4 - foo
> 1 | Hello, world
         ^~",
  ],
  [
    "Error: ./bar: bar",
  ],
]
`;

exports[`TerminalLogger POSIX typed errors 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX typed errors 2`] = `
[
  [
    "Error: Foo!",
  ],
  [
    "Error: Bar!",
  ],
]
`;

exports[`TerminalLogger POSIX typed errors for source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX typed errors for source 2`] = `
[
  [
    "Error: ./bar:1:4 - Foo!
> 1 | Hello, world
         ^~",
  ],
  [
    "Error: ./bar: Bar!",
  ],
]
`;

exports[`TerminalLogger POSIX uncaught error 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX uncaught error 2`] = `
[
  [
    "Internal compiler error: Unhandled: Error: Uncaught!
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger POSIX uncaught error in source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger POSIX uncaught error in source 2`] = `
[
  [
    "Internal compiler error: Unhandled: Error: hehe
while compiling /foo/bar
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger POSIX warn verbosity does not show info 1`] = `undefined`;

exports[`TerminalLogger POSIX warn verbosity does not show info 2`] = `
[
  [
    "Warning: Warn",
  ],
]
`;

exports[`TerminalLogger Windows all info logs are logged 1`] = `undefined`;

exports[`TerminalLogger Windows all info logs are logged 2`] = `
[
  [
    "Info 1",
  ],
  [
    "Info 2",
  ],
]
`;

exports[`TerminalLogger Windows exit on error 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows exit on error 2`] = `
[
  [
    "Error: foo",
  ],
  [
    "Error: bar",
  ],
]
`;

exports[`TerminalLogger Windows internal error in source at range 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows internal error in source at range 2`] = `
[
  [
    "Internal compiler error: .\\bar:1:4 - OMG
> 1 | Hello, world
         ^~
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger Windows logger internal error 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows logger internal error 2`] = `
[
  [
    "Internal compiler error: OMG
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger Windows logger internal error in source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows logger internal error in source 2`] = `
[
  [
    "Internal compiler error: .\\bar: OMG
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger Windows multiple errors 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows multiple errors 2`] = `
[
  [
    "Error: foo",
  ],
  [
    "Error: bar",
  ],
]
`;

exports[`TerminalLogger Windows multiple errors inside source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows multiple errors inside source 2`] = `
[
  [
    "Error: .\\bar:1:4 - foo
> 1 | Hello, world
         ^~",
  ],
  [
    "Error: .\\bar: bar",
  ],
]
`;

exports[`TerminalLogger Windows only first error without error recovery 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows only first error without error recovery 2`] = `
[
  [
    "Error: Error 1",
  ],
]
`;

exports[`TerminalLogger Windows path is resolved relative to cwd 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows path is resolved relative to cwd 2`] = `
[
  [
    "Error: See .\\bar.",
  ],
]
`;

exports[`TerminalLogger Windows raw internal error 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows raw internal error 2`] = `
[
  [
    "Internal compiler error: OMG
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger Windows raw internal error in source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows raw internal error in source 2`] = `
[
  [
    "Internal compiler error: OMG
while compiling /foo/bar
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger Windows source inside multiple errors 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows source inside multiple errors 2`] = `
[
  [
    "Error: .\\bar:1:4 - foo
> 1 | Hello, world
         ^~",
  ],
  [
    "Error: .\\bar: bar",
  ],
]
`;

exports[`TerminalLogger Windows typed errors 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows typed errors 2`] = `
[
  [
    "Error: Foo!",
  ],
  [
    "Error: Bar!",
  ],
]
`;

exports[`TerminalLogger Windows typed errors for source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows typed errors for source 2`] = `
[
  [
    "Error: .\\bar:1:4 - Foo!
> 1 | Hello, world
         ^~",
  ],
  [
    "Error: .\\bar: Bar!",
  ],
]
`;

exports[`TerminalLogger Windows uncaught error 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows uncaught error 2`] = `
[
  [
    "Internal compiler error: Unhandled: Error: Uncaught!
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger Windows uncaught error in source 1`] = `"process.exit called with code 30"`;

exports[`TerminalLogger Windows uncaught error in source 2`] = `
[
  [
    "Internal compiler error: Unhandled: Error: hehe
while compiling /foo/bar
Please report at https://github.com/tact-lang/tact/issues",
  ],
]
`;

exports[`TerminalLogger Windows warn verbosity does not show info 1`] = `undefined`;

exports[`TerminalLogger Windows warn verbosity does not show info 2`] = `
[
  [
    "Warning: Warn",
  ],
]
`;
